### PR TITLE
Single RF loading at training and dl1->dl2

### PR DESCRIPTION
**Pull request by gabemery**
To solve #979 and reduce the memory usage during training and usage of random forest.

- [x] Free space in build_model()
- [x] Change apply_model or script logic to not require to pre-load RF
- [x] Finalise

### CODE REVIEW

File: lstchain/reco/dl1_to_dl2.py [hunk 1]
```diff
@@ -402,11 +403,33 @@ def build_models(filegammas, fileprotons,
 
     reg_energy = train_energy(df_gamma_reg, custom_config=config)
 
+    if save_models:
+        os.makedirs(path_models, exist_ok=True)
+
+        file_reg_energy = path_models + "/reg_energy.sav"
+        joblib.dump(reg_energy, file_reg_energy, compress=3)
+    if free_model_memory:
```
Comment (maxnoe) [inline lstchain/reco/dl1_to_dl2.py:411]: There is not really a reason to put this behind a flag, is there? The variable is not used after dumping.

Comment (gabemery) [reply lstchain/reco/dl1_to_dl2.py:411]: For some reason the function returns the models (probably useful if not saved). I personally don't know if there are any use case. If not we could remove save_models and free_model_memory

Comment (gabemery) [reply lstchain/reco/dl1_to_dl2.py:411]: It is at least used in test currently

File: lstchain/reco/dl1_to_dl2.py [hunk 2]
```diff
@@ -527,19 +528,21 @@ def apply_models(dl1,
...
+    file_reg_disp_vector: string or Random Forest Regressor
+        Location of the, or pre-loaded, RF for disp vector reconstruction
+    file_reg_disp_norm: string or Random Forest Regressor
+        Location of the, or pre-loaded, RF for disp norm reconstruction
+    file_cls_disp_sign: string or Random Forest Classifier
+        Location of the, or pre-loaded, RF for disp sign reconstruction
     focal_length: `astropy.unit`
     custom_config: dictionnary
         Modified configuration to update the standard one
+    pre_loaded: bool
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply lstchain/reco/dl1_to_dl2.py:544]: Calling something that could be a loaded model a `file_classifier` is very misleading.

[the remark this answers is not in the document]
Comment (gabemery) [reply lstchain/reco/dl1_to_dl2.py:544]: I also don't want to unload the models if they are loaded outside of the function so a flag is needed either as argument or dynamically in the function

[the remark this answers is not in the document]
Comment (maxnoe) [reply lstchain/reco/dl1_to_dl2.py:544]: That cannot happen in python. Calling `del` inside the function will remove the reference from inside the function. The memory will only be freed up of the reference count reaches zero. 

`del` does not delete the object. It removes the name pointing to the object, decrementing the reference count.

[the remark this answers is not in the document]
Comment (maxnoe) [reply lstchain/reco/dl1_to_dl2.py:544]: This:

```
class Model:
    def __del__(self):
        print('Model.__del__')



def foo(model):
    print('foo begin')
    del model
    print('foo end')



def main():
    print('main begin')
    model = Model()
    foo(model)
    print('main end')


if __name__ == '__main__':
    main()


```

prints:

```
main begin
foo begin
foo end
main end
Model.__del__
```

[the remark this answers is not in the document]
Comment (gabemery) [reply lstchain/reco/dl1_to_dl2.py:544]: Ok I will update this then

File: lstchain/reco/dl1_to_dl2.py [hunk 3]
```diff
@@ -523,18 +530,18 @@ def apply_models(dl1,
...
-    reg_energy: Random Forest Regressor
-        RF for Energy reconstruction
-    reg_disp_vector: Random Forest Regressor
-        RF for disp vector reconstruction
-    reg_disp_norm: Random Forest Regressor
-        RF for disp norm reconstruction
-    cls_disp_sign: Random Forest Classifier
-        RF for disp sign reconstruction
+    classifier: string or Random Forest Classifier
+        Location of the, or pre-loaded, RF for Gamma/Hadron separation
```
Comment (vuillaut) [inline lstchain/reco/dl1_to_dl2.py:534]: unclear docstring to me.
type should be string | Path | bytes | `sklearn.ensemble.RandomForestClassifier`
and path to the random forest filename or file or pre-loaded RandomForestClassifier object as a description.

**Comment by vuillaut** [comment]
If you can just add a test for the new option, otherwise lgtm.

**Review by vuillaut — APPROVED** [review]

**Comment by gabemery** [comment]
Thanks @vuillaut. I am not allowed to merge myself so if anyone with merging right can do it would be nice. Else I will ask on slack later.

**Pull request merged**